src/commands/plugins.rs

```diff
@@ -0,0 +1,275 @@
+#[derive(Subcommand, Debug)]
+pub enum PluginCommands {
+    /// Install plugin from a manifest.
+    ///
+    /// The binary file and manifest of the plugin is copied to the local Spin
+    /// plugins directory.
+    Install(Install),
+
+    /// Remove a plugin from your installation.
+    Uninstall(Uninstall),
+
+    /// Upgrade one or all plugins.
+    Upgrade(Upgrade),
+}
+
+impl PluginCommands {
+    pub async fn run(self) -> Result<()> {
+        match self {
+            PluginCommands::Install(cmd) => cmd.run().await,
+            PluginCommands::Uninstall(cmd) => cmd.run().await,
+            PluginCommands::Upgrade(cmd) => cmd.run().await,
+        }
+    }
+}
+
+/// Install plugins from remote source
+#[derive(Parser, Debug)]
+pub struct Install {
+    /// Name of Spin plugin.
+    #[clap(
+        name = "PLUGIN_NAME",
+        conflicts_with = "REMOTE_PLUGIN_MANIFEST",
+        conflicts_with = "LOCAL_PLUGIN_MANIFEST",
+        required_unless_present_any = ["REMOTE_PLUGIN_MANIFEST", "LOCAL_PLUGIN_MANIFEST"],
+    )]
+    pub name: Option<String>,
+
+    /// Path to local plugin manifest.
+    #[clap(
+        name = "LOCAL_PLUGIN_MANIFEST",
+        short = 'f',
+        long = "file",
+        conflicts_with = "REMOTE_PLUGIN_MANIFEST",
+        conflicts_with = "PLUGIN_NAME"
+    )]
+    pub local_manifest_src: Option<PathBuf>,
+
+    /// Path to remote plugin manifest.
+    #[clap(
+        name = "REMOTE_PLUGIN_MANIFEST",
+        short = 'u',
+        long = "url",
+        conflicts_with = "LOCAL_PLUGIN_MANIFEST",
+        conflicts_with = "PLUGIN_NAME"
+    )]
+    pub remote_manifest_src: Option<Url>,
+
+    /// Skips prompt to accept the installation of the plugin.
+    #[clap(short = 'y', long = "yes", takes_value = false)]
+    pub yes_to_all: bool,
+
+    /// Specific version of a plugin to be install from the centralized plugins
+    /// repository.
+    #[clap(
+        long = "version",
+        short = 'v',
+        conflicts_with = "REMOTE_PLUGIN_MANIFEST",
+        conflicts_with = "LOCAL_PLUGIN_MANIFEST",
+        requires("PLUGIN_NAME")
+    )]
+    pub version: Option<Version>,
+}
+
+impl Install {
+    pub async fn run(self) -> Result<()> {
+        let manifest_location = match (self.local_manifest_src, self.remote_manifest_src, self.name) {
+            (Some(path), None, None) => ManifestLocation::Local(path),
+            (None, Some(url), None) => ManifestLocation::Remote(url),
+            (None, None, Some(name)) => ManifestLocation::PluginsRepository(PluginInfo::new(&name, Url::parse(SPIN_PLUGINS_REPO)?, self.version)),
+            _ => return Err(anyhow::anyhow!("Must provide plugin name for plugin look up xor remote xor local path to plugin manifest")),
+        };
+        PluginInstaller::new(
+            manifest_location,
+            get_spin_plugins_directory()?,
+            self.yes_to_all,
+            env!("VERGEN_BUILD_SEMVER"),
+        )
+        .install()
+        .await?;
```

Comment:
Again, I'm a little surprised at the indentation here.  I thought a vertical chain of methods was usually indented from the code around it.

Reply:
Fair enough, sorry for the red herring.  Just another reason for me to rant about `cargo fmt`.